calyx/opt/src/passes_experimental/fsm_builder.rs

```diff
@@ -0,0 +1,276 @@
+use calyx_ir::{self as ir, GetAttributes};
+use calyx_utils::CalyxResult;
+const ACYCLIC: ir::Attribute =
+    ir::Attribute::Internal(ir::InternalAttr::ACYCLIC);
+
+pub struct FSMBuilder {}
+
+impl Named for FSMBuilder {
+    fn name() -> &'static str {
+        "fsm-builder"
+    }
+    fn description() -> &'static str {
+        "generates medium fsms in one pass for static and dynamic"
+    }
+}
+
+impl ConstructVisitor for FSMBuilder {
+    fn from(_ctx: &ir::Context) -> CalyxResult<Self> {
+        Ok(FSMBuilder {})
+    }
+    fn clear_data(&mut self) {}
+}
+
+// node that passes into this function is a part of the schedule
+impl StaticSchedule<'_, '_> {
+    fn build_abstract(
+        &mut self,
+        scon: &ir::StaticControl,
+        guard: ir::Guard<ir::Nothing>,
+        mut transitions_to_curr: Vec<IncompleteTransition>,
+        looped_once_guard: Option<ir::Guard<ir::Nothing>>,
+    ) -> (Vec<IncompleteTransition>, Option<ir::Guard<ir::Nothing>>) {
+        match scon {
+            ir::StaticControl::Empty(_) => (transitions_to_curr, None),
+            ir::StaticControl::Enable(sen) => {
+                if matches!(sen.get_attributes().get(ACYCLIC), Some(1)) {
+                    // allocate one state per cycle
+                    // for all parts of the FSM that want to transition to this enable,
+                    // register their transitions in self.state2trans
+                    self.register_transitions(
+                        self.state,
+                        &mut transitions_to_curr,
+                        guard.clone(),
+                    );
+
+                    sen.group.borrow().assignments.iter().for_each(|sassign| {
+                        sassign
+                            .guard
+                            .compute_live_states(sen.group.borrow().latency)
+                            .into_iter()
+                            .for_each(|offset| {
+                                // convert the static assignment to a normal one
+                                let mut assign: ir::Assignment<ir::Nothing> =
+                                    ir::Assignment::from(sassign.clone());
+                                // "and" the assignment's guard with argument guard
+                                assign.and_guard(guard.clone());
+                                // add this assignment to the list of assignments
+                                // that are supposed to be valid at this state
+                                self.state2assigns
+                                    .entry(self.state + offset)
+                                    .and_modify(|other_assigns| {
+                                        other_assigns.push(assign.clone())
+                                    })
+                                    .or_insert(vec![assign]);
+                            })
+                    });
+                    self.state += sen.group.borrow().latency;
```

Comment:
The idea here seems to be: since we're in "acyclic" mode, we allocate N states to make N cycles elapse. Maybe a comment would be helpful.